app/services/proxy_status.rb

```diff
@@ -0,0 +1,55 @@
+class ProxyStatus
+  CONNECTION_ERRORS = [Errno::EINVAL, Errno::ECONNRESET, EOFError, Timeout::Error, Errno::ENOENT,
+                       Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError]
+
+  def initialize(proxy, opts = {})
+    raise ::Foreman::Exception.new(N_('Must specify a Smart Proxy to use')) unless proxy.is_a?(SmartProxy)
+    @proxy = proxy
+    @cache_duration = opts[:cache_duration] || 3.minutes
+  end
+
+  def proxy_version_api
+    @proxy_version_api ||= ProxyAPI::Version.new(:url => proxy.url)
+  end
+
+  def proxy_versions
+    Rails.cache.fetch(cache_key, :expires_in => cache_duration) do
+      fetch_proxy_data do
+        proxy_version_api.proxy_versions
+      end
+    end
+  end
+
+  # TODO: extract to another status implementation
+  def tftp_server
+    raise ::Foreman::Exception.new(N_('No TFTP feature for %s') % proxy.to_label) unless proxy.has_feature?('TFTP')
+    Rails.cache.fetch("proxy_#{proxy.id}/tftp_server", :expires_in => cache_duration) do
+      fetch_proxy_data do
+        ProxyAPI::TFTP.new(:url => proxy.url).bootServer
+      end
+    end
+  end
+
+  def revoke_cache!
+    # As memcached does not support delete_matched, we need to delete each
+    Rails.cache.delete(cache_key)
+    Rails.cache.delete("proxy_#{proxy.id}/tftp_server")
+  end
+
+  def cache_key
+    "proxy_#{proxy.id}/versions"
+  end
+
+  private
+
+  attr_reader :proxy, :cache_duration, :api
+  alias_method :version, :proxy_versions
+
+  def fetch_proxy_data &block
```

Comment:
&block isn't necessary here as an argument and makes it slower, `yield` is enough